### ISSUE TITLE
WCAG20-TECHS ARIAパート コメント(3) - ARIA2 HTMLのrequired属性を使用せずaria-requiredプロパティーのみ記述するのは妥当か？

### ISSUE THREAD

**Issue by momdo** [momdo/momdo.github.io#14]
http://momdo.github.io/WCAG20-TECHS_aria.html#ARIA2

少なくともHTML5.1でrequiredを持つ属性はデフォルトの暗黙ARIAセマンティック`aria-required="true"`を持ち、仕様上はHTMLの`required`と`aria-required="true"`を重複して記述すべきでない。
ARIA2における例示は、`aria-required="true"`のみしか記述しておらず、ホスト言語であるHTMLの`required`属性を使用しないのがそもそもおかしいのではないか。

支援技術の実装の観点からどうなのかは未確認。

**Comment by momdo** [comment]
@milk54 さんからNVDA/PC-Talkerの試験結果をいただいた。
https://github.com/milk54/a11yform/blob/master/milk/milkform-reading.txt#L124
これを見る限り、両実装では`required`のみで特に問題なさそう。

**Comment by milk54** [comment]
VoiceOverとTalkBackの読み上げ結果を追記しました。
https://github.com/milk54/a11yform/commit/0b120589c741db4e9e5e78400961c3006ebebb2d?_pjax=%23js-repo-pjax-container
TalkBackではrequired属性のみでは必須と伝わらないのですが、シェア率を考慮するとrequired属性のみでも良いのではないかと思うのですがいかがでしょうか？

**Comment by momdo** [comment]
追加試験ありがとうございます。
2つのメジャーな実装で良好なら、`required`単独でよいのかな、と。（また、HTML5の新属性なのかでもわりと認知されていると個人的に思っています。）

**Comment by milk54** [comment]
返答いただきありがとうございます。
そうですね。私も同じように(<code>required</code>単独で)思います。

**Comment by momdo** [comment]
HTML5は、required属性を定義している。
The HTML5 spec defines required attribute.
http://momdo.github.io/html5/forms.html#the-required-attribute
http://www.w3.org/TR/2014/REC-html5-20141028/forms.html#the-required-attribute

WAI-ARIA 1.1仕様の7.5節によれば、「同一のロールのセマンティックと値をもつホスト言語における機能が使用可能である、かつ著者がホスト言語の機能を使用しないようにするための説得力のある理由がない場合、著者は、WAI-ARIAをもつ他の要素を再利用するよりもむしろ、ホスト言語の機能を使用すべきである。」
Section 7.5 of the WAI-ARIA 1.1 spec says "When a feature in the host language with identical role semantics and values is available, and the author has no compelling reason to avoid using the host language feature, authors SHOULD use the host language features rather than repurpose other elements with WAI-ARIA".
http://momdo.github.io/wai-aria-1.1/#host_general_conflict
http://www.w3.org/TR/wai-aria-1.1/#host_general_conflict

従って、少なくとも例1-3は不適当です。
At least, therefore, example 1-3 of ARIA2 tequniques are not appropriate.

**Comment by makotom** [comment]
横槍失礼.

機械的アクセシビリティ・ユーザーアクセシビリティの双方の観点から, 必須入力項目の機械的明示は大変重要な要件だと考えられます. したがって, 方向性としては, まずは "HTML5 の required との整合性が ARIA in HTML と WAI-ARIA の観点から合っていない" という指摘を入れて, 最終的には必須項目を機械的に明示するという要件の下, 2種類のアプローチ (required 属性を使用できる要素で required 属性から implicit semantics 経由で明示する方法と, required 属性を使用できない要素で aria-required 属性を用いて明示する方法) を挙げる, というところに帰着させるのがよいのではないかと思います.

プロポーザルとしてはこんな感じでしょうか.

ARIA2: Conflict with other ARIA-related specs

Current ARIA2 in WCAG20 showcases the use of the aria-required property to indicate requirement of field input [1].
However, HTML5 has already defined the required attribute, by which the same objective should be achieved [2].

According to WAI-ARIA 1.1, authors should use features provided by HTML, in stead of ones from WAI-ARIA, as long as " a feature in the host language with identical role semantics and values is available, and the author has no compelling reason to avoid using the host language feature" [2]. Considering that the required HTML attribute and the aria-required property are considered to be identical [4] (and is also is recommended to prefer the HTML one to the ARIA one where HTML ones are applicable [5]), examples in ARIA2 conflict with other related specifications.

The idea itself to mandate the indication of input requirements is favourable from the view point of accessibility for users (and, interestingly, for machines, too). Also for the philosophy of the idea, modification on this section is required to make the guideline conformant with existing specifications.

[1] http://www.w3.org/WAI/GL/2016/WD-WCAG20-TECHS-20160105/ARIA2.html
[2] https://www.w3.org/TR/html5/forms.html#the-required-attribute
[3] http://www.w3.org/TR/wai-aria-1.1/#host_general_conflict
[4] https://www.w3.org/TR/html-aria/#attr-required
[5] https://www.w3.org/TR/html-aria/#h-note2

**Comment by momdo** [comment]
@makotom 
ありがとうございます。主にmarkdownをちょこっと修正・加筆してみました。

ARIA2: Conflict with other ARIA-related specs

Current ARIA2 in WCAG20-TECHS showcases the use of the `aria-required` property to indicate requirement of field input [1].
However, HTML5 has already defined the `required` attribute, by which the same objective should be achieved [2].

According to WAI-ARIA 1.1, authors should use features provided by HTML, in stead of ones from WAI-ARIA, as long as "a feature in the host language with identical role semantics and values is available, and the author has no compelling reason to avoid using the host language feature" [2]. Considering that the `required` HTML attribute and the `aria-required` ARIA property are considered to be identical [4] \(and is also is recommended to prefer the HTML one to the ARIA one where HTML ones are applicable [5]), examples in ARIA2 conflict with other related specifications.

The idea itself to mandate the indication of input requirements is favourable from the view point of accessibility for users (and, interestingly, for machines, too). Also for the philosophy of the idea, modification on this section is required to make the guideline conformant with existing specifications.

[1] https://www.w3.org/WAI/GL/2016/WD-WCAG20-TECHS-20160105/ARIA2.html
[2] https://www.w3.org/TR/html5/forms.html#the-required-attribute
[3] https://www.w3.org/TR/wai-aria-1.1/#host_general_conflict
[4] https://www.w3.org/TR/html-aria/#attr-required
[5] https://www.w3.org/TR/html-aria/#h-note2

**Comment by momdo** [comment]
issue 145して報告した。

**Comment by momdo** [comment]
4月15日付けで本家issueがclosedになっていたので閉じる。
